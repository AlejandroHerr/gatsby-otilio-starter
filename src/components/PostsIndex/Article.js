import React from 'react';
import Link from 'gatsby-link';

import { articleType } from '../../propTypes/postsIndex';

import styles from './PostsIndex.module.scss';

const getDisplayDate = date => date.toLocaleDateString('en-us', { day: 'numeric', month: 'short', year: 'numeric' });

const Article = ({ article }) => {
  const {
    author,
    date,
    excerpt,
    slug,
    tags,
    title,
  } = article;

  const postDate = new Date(date);

  return (
    <article className={styles.posts_index_article}>
      <div className={styles.posts_index_article__inner}>
        <header className="post-header">
          <h2 className={styles.posts_index_article__title}>
            <Link
              className={styles.posts_index_article__title__link}
              to={slug}
            >
              {title
              }
            </Link>
          </h2>
          <span className={styles.posts_index_article__meta}>
            {author}
            {tags && ` on ${tags.join(', ')}`}
            {' | '}
            <time dateTime={postDate.toISOString()}>{getDisplayDate(postDate)}</time>
          </span>
          <div className={styles.posts_index_article__clear} />
        </header>
        <section>
          <p className={styles.posts_index_article__excerpt}>{excerpt}</p>
        </section>
      </div>
      <div className={styles.posts_index_article__separator} />
    </article>
  );
};

Article.propTypes = {
  article: articleType.isRequired,
};

export default Article;
