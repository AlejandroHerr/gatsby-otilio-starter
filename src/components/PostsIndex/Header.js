// @flow
import React from 'react';
import classnames from 'classnames';

import type { SizeImageSharpType } from '../../types/gatsby';

import HeroHeader from '../HeroHeader';

import styles from './PostsIndex.module.scss';

type PropsType = {
  description: string,
  image: SizeImageSharpType,
  title: string,
};

const MainHeader = ({
  description, image, title,
}: PropsType) => (
  <HeroHeader
    className={classnames(styles.posts_index_header, {
        [styles['posts_index_header--has_image']]: !!image,
      })}
    image={image}
  >
    <h1
      className={classnames(styles.posts_index_header__title, {
          [styles['posts_index_header__title--has_image']]: !!image,
        })}
    >
      {title}
    </h1>
    <span
      className={classnames(styles.posts_index_header__body, {
          [styles['posts_index_header__body--has_image']]: !!image,
        })}
    >
      {description}
    </span>
  </HeroHeader>
);

export default MainHeader;
