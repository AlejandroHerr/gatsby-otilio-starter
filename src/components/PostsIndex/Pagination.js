// @flow
import React from 'react';
import Link from 'gatsby-link';
import classnames from 'classnames';

import type { PaginationType } from '../../types/postsIndex';

import ByteIcon from '../icons/ByteIcon';

import styles from './PostsIndex.module.scss';

const Pagination = ({ page, pageCount, pathPrefix }: PaginationType) => (
  <nav className={styles.posts_index_pagination}>
    <div className={styles.posts_index_pagination__inner}>
      {page > 1 &&
        <Link
          className={classnames(
            styles.posts_index_pagination__link,
            styles.posts_index_pagination__next,
          )}
          to={page > 2 ? `${pathPrefix}/${page - 1}` : '/'}
        >
          <span className={styles.posts_index_pagination__label}><ByteIcon icon="chevron_left" /> Newer Posts</span>
        </Link>
      }
      <span className={styles.posts_index_pagination__info}>{`Page ${page} of ${pageCount}`}</span>
      {page < pageCount &&
        <Link
          className={classnames(
            styles.posts_index_pagination__link,
            styles.posts_index_pagination__prev,
          )}
          to={`${pathPrefix}/${page + 1}`}
        >
          <span className={styles.posts_index_pagination__label}>Older Posts <ByteIcon icon="chevron_right" /></span>
        </Link>
      }
      <div className={styles.posts_index_pagination__clear} />
      <div className={styles.posts_index_pagination__separator} />
    </div>
  </nav>
);

export default Pagination;
