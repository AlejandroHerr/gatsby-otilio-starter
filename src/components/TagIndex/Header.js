// @flow
import React from 'react';
import classnames from 'classnames';

import type { SizeImageSharpType } from '../../types/gatsby';

import HeroHeader from '../HeroHeader';

import styles from './TagIndex.module.scss';

type PropsType = {
  image?: ?SizeImageSharpType,
  title: string,
  total: number,
};

const MainHeader = ({
  image, title, total,
}: PropsType) => (
  <HeroHeader
    className={classnames(styles.tag_index_header, {
      [styles['tag_index_header--has_image']]: !!image,
    })}
    image={image}
  >
    <h1
      className={classnames(styles.tag_index_header__title, {
          [styles['tag_index_header__title--has_image']]: !!image,
        })}
    >
      {title}
    </h1>
    <span
      className={classnames(styles.tag_index_header__body, {
          [styles['tag_index_header__body--has_image']]: !!image,
        })}
    >
      {`${total} posts in this category`}
    </span>
  </HeroHeader>
);


MainHeader.defaultProps = {
  image: null,
};

export default MainHeader;
