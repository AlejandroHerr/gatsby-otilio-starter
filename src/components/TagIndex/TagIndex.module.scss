@import '../../colors';
@import '../../fonts';

.tag_index_header__title, {
  display: block;
  position: relative;
  margin: 0;
  padding: 0;
  color: $color-black;
  font-size: 2.5em;
  font-weight: 700;
  line-height: 1em;
  text-decoration: none;
  z-index: 10;

  a {
    color: inherit;
    font-weight: inherit;
    text-decoration: none;
  }
}

.tag_index_header__body {
  display: block;
  position: relative;
  margin: .25em 0;
  padding: 0;
  font-size: 1em;
  line-height: 1.5em;
  z-index: 10;
}

.tag_index_header__title--has_image,
.tag_index_header__body--has_image {
  color: $color-white;
}
