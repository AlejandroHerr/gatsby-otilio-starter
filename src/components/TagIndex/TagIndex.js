// @flow
import React from 'react';

import type { PathContextType, SiteInfoType } from '../../types/tagIndex';

import Header from './Header';
import Article from '../PostsIndex/Article';
import Pagination from '../PostsIndex/Pagination';

type PropsType = {
  pathContext: PathContextType,
  siteInfo: SiteInfoType,
};

const TagIndex = ({ siteInfo, pathContext }: PropsType) => {
  const {
    group: posts,
    first,
    last,
    index,
    pageCount,
    pathPrefix,
    additionalContext,
  } = pathContext;

  return (
    <div>
      <Header
        image={additionalContext.cover && additionalContext.cover.sizes}
        title={additionalContext.tag}
        total={additionalContext.total}
      />
      <main role="main">
        {posts.map(article => (
          <Article
            key={article.slug}
            article={article}
          />
        ))}
        <Pagination
          first={first}
          last={last}
          index={index}
          pageCount={pageCount}
          pathPrefix={pathPrefix}
        />
      </main>
    </div>
  );
};

export default TagIndex;
