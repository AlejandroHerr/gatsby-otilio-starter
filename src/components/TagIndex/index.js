// @flow
import TagIndex from './TagIndex.js';

export default TagIndex;
