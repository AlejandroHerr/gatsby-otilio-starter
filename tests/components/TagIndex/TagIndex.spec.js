import React from 'react';
import { shallow } from 'enzyme';

import TagIndex from '../../../src/components/TagIndex/';
import Header from '../../../src/components/TagIndex/Header';
import Article from '../../../src/components/PostsIndex/Article';
import Pagination from '../../../src/components/PostsIndex/Pagination';

const defaultProps = {
  pathContext: {
    group: [{
      slug: 'slug1',
    },
    {
      slug: 'slug2',
    }],
    first: true,
    last: false,
    index: 1,
    pageCount: 2,
    pathPrefix: 'prefix',
    additionalContext: {
      tag: 'Test Tag',
      total: 10,
    },
  },
};

const setup = (props = {}) => {
  const tagIndex = shallow(<TagIndex {...defaultProps} {...props} />);

  return {
    tagIndex,
    header: tagIndex.find(Header),
    article: tagIndex.find(Article),
    pagination: tagIndex.find(Pagination),
  };
};

describe('TagIndex', () => {
  it('should render a TagIndex', () => {
    const {
      tagIndex, header, article, pagination,
    } = setup();

    expect(header).toHaveLength(1);
    expect(article).toHaveLength(defaultProps.pathContext.group.length);
    expect(pagination).toHaveLength(1);
    expect(tagIndex).toMatchSnapshot();
  });
  it('should render a TagIndex with cover', () => {
    const sizes = {
      src: 'testSrc',
    };
    const {
      tagIndex, header, article, pagination,
    } = setup({
      pathContext: {
        ...defaultProps.pathContext,
        additionalContext: {
          ...defaultProps.pathContext.additionalContext,
          cover: {
            sizes,
          },
        },
      },
    });

    expect(header).toHaveLength(1);
    expect(header.prop('image')).toBe(sizes);
    expect(article).toHaveLength(defaultProps.pathContext.group.length);
    expect(pagination).toHaveLength(1);
    expect(tagIndex).toMatchSnapshot();
  });
});
