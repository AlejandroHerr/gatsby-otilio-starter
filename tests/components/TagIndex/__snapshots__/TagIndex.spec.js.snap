// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagIndex should render a TagIndex 1`] = `
<div>
  <MainHeader
    image={null}
    title="Test Tag"
    total={10}
  />
  <main
    role="main"
  >
    <Article
      article={
        Object {
          "slug": "slug1",
        }
      }
      key="slug1"
    />
    <Article
      article={
        Object {
          "slug": "slug2",
        }
      }
      key="slug2"
    />
    <Pagination
      first={true}
      index={1}
      last={false}
      pageCount={2}
      pathPrefix="prefix"
    />
  </main>
</div>
`;

exports[`TagIndex should render a TagIndex with cover 1`] = `
<div>
  <MainHeader
    image={null}
    title="Test Tag"
    total={10}
  />
  <main
    role="main"
  >
    <Article
      article={
        Object {
          "slug": "slug1",
        }
      }
      key="slug1"
    />
    <Article
      article={
        Object {
          "slug": "slug2",
        }
      }
      key="slug2"
    />
    <Pagination
      first={true}
      index={1}
      last={false}
      pageCount={2}
      pathPrefix="prefix"
    />
  </main>
</div>
`;
